src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,20 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
     res.refs = repeat(res.refs, inner = [v.inner], outer = [v.outer])
     res
 end
+
+
+transpose1(df::AbstractDataFrame, indexcol=1) = unstack(stack(df,Not(indexcol)), :variable, indexcol,:value)
+
+function transpose2(df::AbstractDataFrame, indexcol=1)
+    m = permutedims(Matrix(df[!, Not(indexcol)]))
```

Comment:
here you can use `typejoin` to find an appropriate type.

Reply:
`promote` is probably even better as it will take care of choosing the best numeric types if numeric columns have different types. It will also return `Union{T,Missing}` rather than `Any`.

Reply:
Actually `Matrix(::DataFrame)` already uses `promote_type`. AFAICT the behavior is correct. You get an `Any` column only because one of the input columns is `Anÿ`, but that's expected.

Reply:
I proposed `typejoin` so that taking transpose of a transpose does not change values and gets back where we started. But maybe it is OK to make a promotion (for a `Matrix` for sure it makes sense, but I was not sure here)

Reply:
Ah yes. But getting concrete types sounds more important?

Reply:
Let us discuss it, as probably most important problematic case is `Bool` (as you later might want to use it to check logical conditions):
```
julia> df = DataFrame([[true,false],1:2,1.0:2.0])
2×3 DataFrame
│ Row │ x1   │ x2    │ x3      │
│     │ Bool │ Int64 │ Float64 │
├─────┼──────┼───────┼─────────┤
│ 1   │ 1    │ 1     │ 1.0     │
│ 2   │ 0    │ 2     │ 2.0     │

julia> Matrix(df)
2×3 Array{Float64,2}:
 1.0  1.0  1.0
 0.0  2.0  2.0
```